code/modules/station_goals/bluespace_tap.dm

```diff
@@ -331,120 +320,97 @@
-  * Note that this is not necessarily the current power use.
-  * * i_level - The hypothetical input level for which we want to know the power use.
-  */
-/obj/machinery/power/bluespace_tap/proc/get_power_use(i_level)
-	if(!i_level)
-		return 0
-	return power_needs[i_level]
+/obj/machinery/power/bluespace_tap/proc/set_power(t_power)
+	desired_mining_power = max(t_power, 0)
+	// Round down to nearest MW if above 1 MW
+	if(desired_mining_power > 1 MW)
+		desired_mining_power = desired_mining_power - (desired_mining_power % (1 MW))
 
 /obj/machinery/power/bluespace_tap/process()
-	actual_power_usage = get_power_use(input_level)
-	if(get_surplus() < actual_power_usage)	//not enough power, so turn down a level
-		input_level--
-		update_icon()
-		return	// and no mining gets done
-	if(actual_power_usage)
-		consume_direct_power(actual_power_usage)
-		//2 points per level up to level 10 and 4 points per MW (or 5 when emmaged).
-		var/points_to_add = min(base_points * 10, base_points * input_level) + actual_power_usage * (conversion_ratio + emagged * 1e-6)
-		points += points_to_add	// point generation, emagging gets you 'free' points at the cost of higher anomaly chance
-		total_points += points_to_add
-	// Between levels 15 and 18 get one level of overhead per 5MW of surplus power. Above level 18 get 1 level per 10MW of surplus power.
-	overhead = input_level >= 18  ? get_surplus() * 1e-7 : get_surplus() * 2e-7
-	stabilizer_power = stabilizers && input_level > 15 ? input_level >= 18 ? min(get_surplus() , (input_level - 15) * 1e7) : min(get_surplus() , (input_level - 15) * 0.5e7) : 0
-	consume_direct_power(stabilizer_power)
-	// actual input level changes slowly
-	// holy shit every proccess this
-	if(input_level < desired_level && (get_surplus() + get_power_use(input_level) + stabilizer_power >= get_power_use(input_level + 1)))
-		input_level++
-		update_icon()
-	else if(input_level > desired_level)
-		input_level--
-		update_icon()
-	// Stabilizers reduce the chance of portals. prob with values less than 0 treat it as 0.
-	if(prob(input_level - (safe_levels + stabilizers * overhead) + (emagged * 5)))
+	// Power we could mine with (Watts)
+	mining_power = get_surplus()
+
+	// Round down to the nearest MW if above a MW
+	if(mining_power > 1 MW)
+		mining_power = mining_power - mining_power % (1 MW)
+
+	// Always try to use all available power for mining if emagged
+	if(emagged)
+		desired_mining_power = mining_power
+
+	// Stabilizers activate above 15MW of mining power
+	// Stabilizers consume up to 1MW for each 1MW of mining power, consuming less between 15 and 30MW of mining power
+	// If stabilizers have priority they will always consume enough power to stabilize the BSH, limiting mining
+	// emagging disables stabilizers
+	if(stabilizer_priority)
+		stabilizer_power = clamp(desired_mining_power - clamp(30 MW - desired_mining_power, 0, 15 MW), 0, mining_power / 2) * (stabilizers && !emagged)
```

Comment:
Do you think you could break these calculations up into a few lines so it's a little clearer what's going on?

Reply:
I'll try. I'll explain it in a comment anyhow

Reply:
Byond complains when I break it into different lines. I added a comment for each calculation, and there is additional clarification above

Reply:
Done